ENH ColumnTransformer.get_feature_names() handles passthrough

lrjball opened the pull request:
Currently, if ColumnTransformer is called with remainder='passthrough', then get_feature_names() will raise a NotImplementedError, but this pull request adds in that functionality. 

In this pull request, if the ColumnTransformer was fitted on a DataFrame and remainder='passthrough' then the columns which passed through will appear in get_feature_names() as their column names in the DataFrame, and if it was not fitted on a DataFrame, then it will be the indices of the columns which will appear in get_feature_names().

Also, if someone explicitly passes a transformer as the text 'passthrough', then the feature names will be name__{column name}, where column_name is whatever have defined when they passed it.
e.g.
`ct = ColumnTransformer([('trans', 'passthrough', ['col0', 'col1'])])`
will produce features 
`['trans__col0', 'trans__col1']`
which is in keeping with the existing behavior for other transformers.

The behaviour for when a transformer does not have a get_feature_names method has also been changed. Now, instead of an error being raised, the feature names will be given as 'name_x0', ..., 'name_xN'. This allows get_feature_names to be useful by giving at least some indication of where the features came from, even if some of the transformers do not give explicit feature names.

Don't believe this fixes any currently open issues.

lrjball committed:
Implemented get_feature_names() for ColumnTransformer to include 'passthough'

Currently, if remainder='passthrough', then get_feature_names() will raise a NotImplementedError, but this pull request adds in that functionality. Now if the transformer is fit on a DataFrame, then the passthrough columns will appear in gte_feature_names() as the respective column names in the DataFrame, and if it is not a DataFrame then the column indices will be used instead.

jnothman commented on sklearn/compose/tests/test_column_transformer.py:
```diff
@@ -249,6 +249,14 @@ def transform(self, X, y=None):
     assert ct.transformers_[-1][1] == 'drop'
     assert_array_equal(ct.transformers_[-1][2], [1])
 
+    assert_array_equal(ct.fit_transform(X_df2), X_res_first)
```
This is duplicated from above

lrjball replied:
The above lines created df2 with integer column names, but never used it in the test (which I assumed was a typo). So I have duplicated the above tests using df2 instead of df instead of replacing df with df2 just incase it wasn't a typo.

jorisvandenbossche replied:
That indeed looks like a typo, but I think you can rather correct the code above, than duplicating it (since it is not related to the rest of this PR, also feel free to open a separate PR for this)

lrjball replied:
Okay, I will remove the duplication and submit this as a separate pull request, thanks

lrjball replied:
Raised #14128 for this instead

jnothman commented on sklearn/compose/tests/test_column_transformer.py:
```diff
@@ -649,25 +657,53 @@ def test_column_transformer_get_feature_names():
-    assert_raise_message(
-        NotImplementedError, 'get_feature_names is not yet supported',
-        ct.get_feature_names)
+    assert_equal(ct.get_feature_names(), ['trans__a', 'trans__b', 1])
```
Not entirely happy with the mix of string and numeric types.

lrjball replied:
I know what you mean. The alternative would be to cast the array indices to string, would that be better?

jorisvandenbossche replied:
This comes a bit in the general "get feature names" discussion domain. For arrays, we might give default names like "x0", "x1", ... but that's not fully decided yet.

lrjball replied:
"x0", "x1", ... seems to be used in a few other transformers (e.g. PolynomialFeatures), and is mentioned in scikit-learn/enhancement_proposals#18 as well. Will this be dependent on that SLEP then?

jnothman replied:
That SLEP could certainly inform the choice. I'd be most happy with "x0", "x1", ... here for now.

lrjball replied:
Should the indices be the same as the indices in the input array, or should they be from 0 to the number of passthrough columns? I think that keeping the indices the same as the input would make them easier to interpret, but this goes against the behaviour of something like PolynomialFeatures, or even the behaviour implemented here for transformers with no get_feature_names, where they always start counting from "x0" .

If starting from 0, the strings would probably need to start with something like 'passthrough__' to avoid confusion.

jnothman replied:
Number according to the input features, please.

lrjball replied:
I've added this now, thanks

jnothman commented on sklearn/compose/tests/test_column_transformer.py:
```diff
@@ -649,25 +657,53 @@ def test_column_transformer_get_feature_names():
-    assert_raise_message(
-        NotImplementedError, 'get_feature_names is not yet supported',
-        ct.get_feature_names)
+    assert_equal(ct.get_feature_names(), ['trans__0', 'trans__1'])
```
I'm not convinced that we should be applying a prefix if trans does nothing... Can you imagine a context where this helps? Yes, it avoids naming conflicts...

lrjball replied:
This was just to keep it consistent with the 'name__column' syntax used for when an explicit name is passed with a transformer, if for any reason someone passed transformers=[(name, 'passthrough', [columns])]. This seemed more consistent than dropping the name.

jnothman replied:
Hmmm.... I still am not happy with this prefixing. I'd be interested to hear other opinions, or use-cases.

lrjball replied:
I'm not sure why 'passthrough' is a valid option for a transformer at all, it seems like that could just be handled with remainder='passthrough' and dropping any unused columns before using the ColumnTransformer. But as it can be passed as a transformer with a name, it seems to make sense to use that name on the feature names, like with the other transformers. Having said that, these should probably be 'trans_x0', 'trans_x1' to be consistent with that notation. Happy to change it altogether though if there is a better option.

jnothman replied:
> I'm not sure why 'passthrough' is a valid option for a transformer at all, it seems like that could just be handled with remainder='passthrough' and dropping any unused columns before using the ColumnTransformer.

It is available to allow the user to try disabling some transformations in a parameter search such as grid search.

lrjball replied:
Ah okay, so I suppose in that case you would want those disabled columns to be treated the same way as all of the other passthrough columns. I will make this change.

jnothman reviewed (commented):
I think the only reason we left this unimplemented was to avoid making hard commitments when first implementing the ColumnTransformer. Thanks for the pr

lrjball committed:
Updated get_feature_names() to always return feature names.

While making the changes for ‘passthrough’, it seems sensible to make an additional change to ColumnTransformer.get_feature_names(), to make it always returns something. This implementation adds feature names name__x0, …, name__xN for transformers without a get_feature_names method.

It seems harsh to raise an error if there are some transformers which do have feature names, and even if none of them have a get_feature_names then it is still helpful to know which features came from which transformer.

jorisvandenbossche commented on sklearn/compose/_column_transformer.py:
```diff
@@ -332,19 +336,22 @@ def get_feature_names(self):
-            feature_names.extend([name + "__" + f for f in
-                                  trans.get_feature_names()])
+                feature_names.extend([name + "__x" + str(i)
+                                      for i in range(dim)])
```
So this is also giving names if the transformer has no feature names method? (just to clarify, this is not strictly related to the PR of enabling feature names for passthrough?)

lrjball replied:
Yes that's right, it seems better to give default names than to error out especially if several of the other transformers have feature names. And even if none of them do, at least this let's you know which features came from which transformer. I suppose it isn't strictly related to the passthrough parameter, although is pretty similar. Would it be worth making a separate pull request for this?

jnothman replied:
Hmm... Yes, I don't think we should be doing this. I'd rather tell the user that they have to define get_feature_names in a component transformer than to invent something.

lrjball replied:
Fair enough, I suppose this would have just been a workaround until every Transformer has a get_feature_names() method.

lrjball replied:
I've now removed this change

jorisvandenbossche reviewed (commented):
Thanks for working on this!

lrjball committed:
removed typo fix to raise as separate PR

lrjball committed:
Implemented get_feature_names() for ColumnTransformer to include 'passthough'

Currently, if remainder='passthrough', then get_feature_names() will raise a NotImplementedError, but this pull request adds in that functionality. Now if the transformer is fit on a DataFrame, then the passthrough columns will appear in gte_feature_names() as the respective column names in the DataFrame, and if it is not a DataFrame then the column indices will be used instead, where the feature names will be 'xi' for the ith index.

jnothman commented on sklearn/compose/_column_transformer.py:
```diff
@@ -305,9 +305,13 @@ def _validate_remainder(self, X):
+        remaining_idx = sorted(list(set(range(n_columns)) - set(cols)))
+        if hasattr(X, 'columns'):
+            columns = X.columns
+            self._remainder_names = [columns[idx] for idx in remaining_idx]
```
This code can be changed after #14237 is merged

lrjball committed:
Merge branch 'master' into column_transformer_passthrough

lrjball committed:
replaced depricated assert_equal with assert

lrjball committed:
Updated the behaviour of get_feature_names to treat trans='passthrough' the same way
as remainder='passthrough'.

The behaviour of get_feature_names for passthrough is now the following:
- If fitted on a dataframe, then the columns passed with the trans='passthrough'
will be treated as positional if int (in which case the feature name will be
the column name at that position) or they will be used for the actual feature name.
Any columns in remainder='passthrough' will be appended to the end of feature_names,
as the column names from the fitted dataframe
- If fitted on an array, then the column names for both when trans='passthrough' or
when remainder='passthrough' will be 'xi' for each index value i. In terms of ordering,
the remainder columns will again come after the rest of the feature names.

lrjball commented on sklearn/compose/_column_transformer.py:
```diff
@@ -343,14 +343,20 @@ def get_feature_names(self):
-            elif not hasattr(trans, 'get_feature_names'):
+            if trans == 'passthrough':
+                if isinstance(column, (str, int)):
+                    column = [column]
```
I think test coverage dropped because this has not been tested. This can be tested once #14495 is added, or these 2 lines can be removed if it is not added.

lrjball replied:
Fixed coverage issue, as #14495 was closed

lrjball committed:
removed checks on scalar columns for passthrough

Removed as this was waiting on PR #14495, which did not go ahead.

lrjball committed:
Merge branch 'master' into column_transformer_passthrough

adrinjalali commented:
Opened #16444 to discuss the options related to the remaining question.

adrinjalali commented:
@lrjball thanks for your patience. #16444 resolves the prefixing decision issue, and as you can see there, we'd like to have the feature names not prefixed in either of the cases:
- `remainder=passthrough`
- estimator is `passthrough`

Would you have time to apply the changes?

lrjball commented:
@adrinjalali No problem. Can I just confirm this is the agreed logic before I make any changes:
- when a dataframe is passed, the feature names when either the transformer is `'passthrough'` or `remainder='passthrough'` will just be the column names from the dataframe.
- when an array is passed, the feature names for either `passthough` case will be the string `'xi'` where `i` is the index of that feature in the input array.

In either case, if any of the transformers other than `'passthrough'` (or `'drop'`) don't have a `get_feature_names` attribute, an error will be raised.

Is that right?

adrinjalali commented:
Yes, that's how I think of it.

lrjball committed:
Merge branch 'master' into column_transformer_passthrough

lrjball committed:
Fixed missing import issue

Fixed issue caused by missing import introduced when doing a merge in the browser. _check_key_type has been replaced with _determine_key_type.

lrjball commented:
Okay, the behavior is now as agreed above, and the merge conflict has been fixed.

adrinjalali reviewed (approved):
Nice! Thanks for your patience and persistence @lrjball

thomasjpfan commented on sklearn/compose/tests/test_column_transformer.py:
```diff
@@ -668,25 +668,54 @@ def test_column_transformer_get_feature_names():
+    assert ct.get_feature_names() == ['x1', 'x0']
+
+    # passthough transformer with a dataframe
+    pd = pytest.importorskip('pandas')
```
This and onward needs to be its own test otherwise the whole test will be marked as "skip" when pandas is not installed.

lrjball replied:
I had assumed it would only skip the rest of the test and treat it as a pass, but you are right. In that case I will raise a PR for one of the other tests (`test_column_transformer_callable_specifier`) which is also using it part way through the test

thomasjpfan commented on sklearn/compose/_column_transformer.py:
```diff
@@ -540,6 +543,7 @@ def fit_transform(self, X, y=None):
 
         self._update_fitted_transformers(transformers)
         self._validate_output(Xs)
+        self._output_dims = [X.shape[1] for X in Xs]
```
Is this used?

lrjball replied:
Thanks for spotting this. It was used in the original PR for a feature which was since removed. I'll update it now

thomasjpfan commented on sklearn/compose/_column_transformer.py:
```diff
@@ -313,19 +313,18 @@ def _validate_remainder(self, X):
-        remaining_idx = sorted(remaining_idx) or None
 
-        self._remainder = ('remainder', self.remainder, remaining_idx)
+        remaining_idx = sorted(list(set(range(self._n_features)) - set(cols)))
```
```suggestion
        remaining_idx = sorted(set(range(self._n_features)) - set(cols))
```

thomasjpfan reviewed (commented):
Thank you for the PR @lrjball !

lrjball committed:
Update sklearn/compose/_column_transformer.py

Co-Authored-By: Thomas J Fan <thomasjpfan@gmail.com>

lrjball committed:
Separated pandas test into own function, and removed unused attribute.

- Seperated the pandas part of the test into its own function to avoid the whole test being skipped when pandas is not installed.
- Removed the unused _output_dims attribute.

lrjball committed:
Merge branch 'column_transformer_passthrough' of https://github.com/lrjball/scikit-learn into column_transformer_passthrough

jnothman reviewed (approved):
Please add an `|Enhancement|` entry to the change log at `doc/whats_new/v0.23.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`

lrjball committed:
added enhancement entry to whats_new

lrjball commented:
Thanks, I have added an entry to the change log now. Is it worth me updating the get_feature_names docstring as well, or is that unnecessary?

lrjball committed:
Added support for boolean masks and slices

Added support for mask and slices for both dataframes and arrays, as well as tests for each case.

lrjball commented on sklearn/compose/tests/test_column_transformer.py:
```diff
@@ -668,25 +668,58 @@ def test_column_transformer_get_feature_names():
+    ct.fit(X_df)
+    assert ct.get_feature_names() == ['col0__a', 'col0__b', 'col1']
+
+    ct = ColumnTransformer([('trans', 'passthrough', ['col1'])],
```
Good point, it didn't. I've updated it to support boolean masks as well as slices for both dataframes and arrays. I've added tests for each as well.
